app/models/vocabulary.rb

```diff
@@ -1,15 +1,77 @@
 class Vocabulary < ApplicationRecord
+  ALPHANUMERIC_UNDERSCORE_KEY_REGEX = /\A[a-z]+[a-z0-9_]*\z/
+  RESERVED_FIELD_NAMES = %w(
+    pref_label alt_label uri uri_hash vocabulary vocabulary_id,
+    authority term_type custom_fields uuid created_at updated_at
+  ).freeze
+  DATA_TYPES = %w(string number boolean).freeze
+
   has_many :terms, dependent: :destroy
 
   validates :string_key, presence: true, uniqueness: true, format: {
-    with: /\A[a-z]+[a-z0-9_]*\z/,
+    with: ALPHANUMERIC_UNDERSCORE_KEY_REGEX,
     message: 'only allows lowercase alphanumeric characters and underscores'
   }
   validates :label,      presence: true
+  validate :validate_custom_fields
 
   store :custom_fields, coder: JSON
 
   def to_api
     as_json(only: [:string_key, :label, :custom_fields])
   end
+
+  def add_custom_field(options = {})
+    field_key = options[:field_key]
+
+    if field_key.blank?
+      raise 'field_key cannot be blank'
+    elsif custom_fields[field_key].present?
+      raise 'field_key cannot be added because its already a custom field'
+    else
+      custom_fields[field_key] = { data_type: options[:data_type], label: options[:label] }
+    end
+  end
+
+  def update_custom_field(options = {})
+    field_key = options[:field_key]
+
+    if field_key.blank?
+      raise 'field_key cannot be blank'
```

Comment:
Same thing here as my earlier comment about changing `raise` errors to errors on the vocabulary object.  They won't ever be shown as part of a normal error response in CustomFieldsController#update.